File: examples/csharp/avalonia/NotesShowcase.Tests/ViewModels/NotebooksRootVMTests.cs
using System.Reactive.Concurrency;
using System.Reactive.Linq;
using NotesShowcase.Models;
using NotesShowcase.ViewModels;
using VMx.Capabilities;
using VMx.Hierarchical;
using VMx.Notifications;
using VMx.Services;
using Xunit;

namespace NotesShowcase.Tests.ViewModels;

public sealed class NotebooksRootVMTests
{
    private static NotebooksRootVM BuildVM(INoteRepository repo)
    {
        var hub = new MessageHub();
        var dispatcher = new RxDispatcher(ImmediateScheduler.Instance, ImmediateScheduler.Instance);
        return NotebooksRootVM.Builder()
            .Name("root")
            .Services(hub, dispatcher)
            .Repository(repo)
            .Build();
    }

    [Fact]
    public void Implements_INewCreatable_and_IReconstructable()
    {
        var repo = new InMemoryNoteRepository(SeedData.Build());
        var vm = BuildVM(repo);
        Assert.IsAssignableFrom<INewCreatable>(vm);
        Assert.IsAssignableFrom<IReconstructable>(vm);
    }

    [Fact]
    public async Task PopulateAsync_loads_seed_notebooks_and_assigns_roots()
    {
        var repo = new InMemoryNoteRepository(
            SeedData.Build(), loadAllDelay: TimeSpan.Zero);
        var vm = BuildVM(repo);
        vm.Construct();
        await vm.PopulateAsync();
        Assert.Equal(5, vm.All.Count);
        // 4 root notebooks per SeedData.
        Assert.Equal(4, vm.Roots.Count);
        // nb-specs is a child of nb-work.
        var work = vm.All.Single(n => n.Model.Id == "nb-work");
        var specs = vm.ChildrenOf(work).Single();
        Assert.Equal("nb-specs", specs.Model.Id);
    }

    [Fact]
    public async Task AddNotebookAsync_emits_TreeStructureChangedMessage_and_appends_child()
    {
        var repo = new InMemoryNoteRepository(
            SeedData.Build(),
            loadAllDelay: TimeSpan.Zero,
            addNotebookDelay: TimeSpan.Zero);
        var vm = BuildVM(repo);
        vm.Construct();
        await vm.PopulateAsync();
        var observed = new List<TreeStructureChangedMessage>();
        using var sub = vm.Hub.Messages
            .OfType<TreeStructureChangedMessage>()
            .Subscribe(observed.Add);

        await vm.AddNotebookAsync(parentId: null, name: "Inbox");

        Assert.NotEmpty(observed);
        Assert.Contains(vm.Walk(), n => n.Model.Name == "Inbox");
    }

    [Fact]
    public async Task Current_setter_round_trip_emits_PropertyChangedMessage()
    {
        var repo = new InMemoryNoteRepository(
            SeedData.Build(), loadAllDelay: TimeSpan.Zero);
        var vm = BuildVM(repo);
        vm.Construct();
        await vm.PopulateAsync();
        var first = vm.Roots[0];
        vm.Current = first;
        Assert.Same(first, vm.Current);
        // Idempotent: setting the same value is a no-op.
        vm.Current = first;
        Assert.Same(first, vm.Current);
        // Clear:
        vm.Current = null;
        Assert.Null(vm.Current);
    }

    [Fact]
    public async Task AddNotebookCommand_executes_when_constructed()
    {
        var repo = new InMemoryNoteRepository(
            SeedData.Build(),
            loadAllDelay: TimeSpan.Zero,
            addNotebookDelay: TimeSpan.Zero);
        var vm = BuildVM(repo);
        vm.Construct();
        await vm.PopulateAsync();
        var before = vm.All.Count;
        vm.AddNotebookCommand.Execute(null);
        // CreateNew uses fire-and-forget — wait for the new notebook to appear.
        await TestWait.WaitUntilAsync(() => vm.All.Count > before);
        Assert.True(vm.All.Count > before);
    }

    // ── Phase 5.a binding gap #2: hierarchical Children accessor ──────────

    [Fact]
    public async Task After_PopulateAsync_each_notebook_resolves_Children_via_parent_id()
    {
        var repo = new InMemoryNoteRepository(SeedData.Build(), loadAllDelay: TimeSpan.Zero);
        var vm = BuildVM(repo);
        vm.Construct();
        await vm.PopulateAsync();

        var work = vm.All.Single(n => n.Model.Id == "nb-work");
        var specs = vm.All.Single(n => n.Model.Id == "nb-specs");

        // "Specs" must appear under "Work" via the Children accessor — this is
        // what the Avalonia TreeView's TreeDataTemplate.ItemsSource binds to.
        Assert.Contains(specs, work.Children);
        Assert.Single(work.Children);
        // Leaf notebooks (no children) return an empty list.
        Assert.Empty(specs.Children);
    }

    // ── notification coverage: "Notebook added" notification ──────────────────

    [Fact]
    public async Task AddNotebookAsync_publishes_Notebook_added_notification_when_hub_wired()
    {
        var repo = new InMemoryNoteRepository(
            SeedData.Build(),
            loadAllDelay: TimeSpan.Zero,
            addNotebookDelay: TimeSpan.Zero);
        var hub = new MessageHub();
        var dispatcher = new RxDispatcher(ImmediateScheduler.Instance, ImmediateScheduler.Instance);
        using var notificationHub = new NotificationHub();
        var observed = new List<Notification>();
        using var sub = notificationHub.Pending.Subscribe(snapshot =>
        {
            foreach (var n in snapshot) if (!observed.Contains(n)) observed.Add(n);
        });
        var vm = NotebooksRootVM.Builder()
            .Name("root").Services(hub, dispatcher).Repository(repo)
            .NotificationHub(notificationHub).Build();
        vm.Construct();
        await vm.PopulateAsync();

        await vm.AddNotebookAsync(parentId: null, name: "Inbox");

        Assert.Contains(observed, n => n.Message.Contains("Notebook added") && n.Message.Contains("Inbox"));
    }

    [Fact]
    public async Task AddNotebookAsync_with_parentId_appears_under_parent_Children()
    {
        var repo = new InMemoryNoteRepository(
            SeedData.Build(),
            loadAllDelay: TimeSpan.Zero,
            addNotebookDelay: TimeSpan.Zero);
        var vm = BuildVM(repo);
        vm.Construct();
        await vm.PopulateAsync();
        var work = vm.All.Single(n => n.Model.Id == "nb-work");
        var beforeCount = work.Children.Count;

        await vm.AddNotebookAsync(parentId: "nb-work", name: "Subspecs");

        var added = vm.All.Single(n => n.Model.Name == "Subspecs");
        Assert.Equal("nb-work", added.Model.ParentId);
        Assert.Contains(added, work.Children);
        Assert.Equal(beforeCount + 1, work.Children.Count);
    }

    [Fact]
    public async Task PopulateAsync_raises_PropertyChanged_for_Roots()
    {
        // Regression: Roots is a computed snapshot; without this raise an
        // already-bound TreeView stays empty forever (the App binds the
        // window BEFORE ConstructAsync completes).
        var repo = new InMemoryNoteRepository(SeedData.Build());
        var vm = BuildVM(repo);
        vm.Construct();
        var raised = new List<string?>();
        ((System.ComponentModel.INotifyPropertyChanged)vm).PropertyChanged +=
            (_, e) => raised.Add(e.PropertyName);

        await vm.PopulateAsync();

        Assert.Contains(nameof(NotebooksRootVM.Roots), raised);
    }

    [Fact]
    public async Task AddNotebookAsync_root_level_raises_PropertyChanged_for_Roots()
    {
        var repo = new InMemoryNoteRepository(SeedData.Build());
        var vm = BuildVM(repo);
        vm.Construct();
        await vm.PopulateAsync();
        var raised = new List<string?>();
        ((System.ComponentModel.INotifyPropertyChanged)vm).PropertyChanged +=
            (_, e) => raised.Add(e.PropertyName);

        await vm.AddNotebookAsync(parentId: null, name: "Fresh Root");

        Assert.Contains(nameof(NotebooksRootVM.Roots), raised);
    }
}

File: examples/swift/notes-showcase/Tests/NotesShowcaseTests/NotebooksRootVMTests.swift
//
// NotebooksRootVMTests — scenario tests for NotebooksRootVM.
//
// Ports NotesShowcase.Tests/ViewModels/NotebooksRootVMTests.cs (C# Avalonia flavor).
// No conformance-ID markers (scenario IDs live in THEME-00x only).
//
import XCTest
import Combine
import VMx
@testable import NotesShowcaseCore

// MARK: - NotebooksRootVMTests

final class NotebooksRootVMTests: XCTestCase {

    // MARK: - Helpers

    private func makeRepo(
        loadAllDelay: TimeInterval = 0,
        addNotebookDelay: TimeInterval = 0
    ) -> InMemoryNoteRepository {
        InMemoryNoteRepository(
            seed: SeedData.build(),
            loadAllDelay: loadAllDelay,
            addNotebookDelay: addNotebookDelay
        )
    }

    private func buildVM(repo: InMemoryNoteRepository) throws -> NotebooksRootVM {
        let hub = MessageHub()
        let dispatcher = ImmediateDispatcher.INSTANCE
        return try NotebooksRootVM.builder()
            .name("root")
            .services(hub: hub, dispatcher: dispatcher)
            .repository(repo)
            .build()
    }

    /// Waits (via cooperative yields) until `condition()` returns `true` or
    /// the attempt limit is exhausted. Used for fire-and-forget async paths.
    private func waitUntil(
        _ condition: @escaping () -> Bool,
        attempts: Int = 50
    ) async {
        for _ in 0..<attempts {
            if condition() { return }
            try? await Task.sleep(nanoseconds: 1_000_000)
        }
    }

    // MARK: - Capability conformances

    func testImplementsNewCreatableAndReconstructable() throws {
        let repo = makeRepo()
        let vm = try buildVM(repo: repo)
        XCTAssertTrue(vm is NewCreatable)
        XCTAssertTrue(vm is Reconstructable)
    }

    // MARK: - populate

    func testPopulateLoadsAllNotebooksAndAssignsRoots() async throws {
        let repo = makeRepo()
        let vm = try buildVM(repo: repo)
        try vm.construct()

        try await vm.populate()

        // SeedData has 5 notebooks total.
        XCTAssertEqual(5, vm.all.count)
        // 4 root notebooks (nb-specs is a child of nb-work).
        XCTAssertEqual(4, vm.roots.count)
        // nb-specs is a child of nb-work via parentId.
        guard let work = vm.all.first(where: { $0.model.id == "nb-work" }) else {
            return XCTFail("nb-work not found after populate")
        }
        let specsChildren = vm.childrenOf(work)
        XCTAssertEqual(1, specsChildren.count)
        XCTAssertEqual("nb-specs", specsChildren[0].model.id)
    }

    // MARK: - addNotebook

    func testAddNotebookEmitsTreeStructureChangedMessageAndAppends() async throws {
        let repo = makeRepo()
        let vm = try buildVM(repo: repo)
        try vm.construct()
        try await vm.populate()

        var observed: [TreeStructureChangedMessage] = []
        var cancellables = Set<AnyCancellable>()
        vm.hub.messages
            .compactMap { $0 as? TreeStructureChangedMessage }
            .sink { observed.append($0) }
            .store(in: &cancellables)

        try await vm.addNotebook(parentId: nil, name: "Inbox")

        XCTAssertFalse(observed.isEmpty, "Expected at least one TreeStructureChangedMessage")
        XCTAssertTrue(vm.walk().contains(where: { $0.model.name == "Inbox" }),
                      "Expected 'Inbox' in walk() after addNotebook")
    }

    // MARK: - current

    func testCurrentSetterRoundTripAndIdempotent() async throws {
        let repo = makeRepo()
        let vm = try buildVM(repo: repo)
        try vm.construct()
        try await vm.populate()

        let first = vm.roots[0]
        vm.current = first
        XCTAssertTrue(vm.current === first)

        // Idempotent: setting the same reference is a no-op.
        vm.current = first
        XCTAssertTrue(vm.current === first)

        // Clear.
        vm.current = nil
        XCTAssertNil(vm.current)
    }

    // MARK: - addNotebookCommand

    func testAddNotebookCommandExecutesWhenConstructed() async throws {
        let repo = makeRepo()
        let vm = try buildVM(repo: repo)
        try vm.construct()
        try await vm.populate()
        let before = vm.all.count

        try await vm.addNotebookCommand.executeAsync()

        XCTAssertTrue(vm.all.count > before,
                      "Expected at least one new notebook after addNotebookCommand")
    }

    // MARK: - Children accessor

    func testAfterPopulateEachNotebookResolvesChildrenViaParentId() async throws {
        let repo = makeRepo()
        let vm = try buildVM(repo: repo)
        try vm.construct()
        try await vm.populate()

        guard let work  = vm.all.first(where: { $0.model.id == "nb-work" }),
              let specs = vm.all.first(where: { $0.model.id == "nb-specs" }) else {
            return XCTFail("nb-work / nb-specs not found")
        }

        // Specs must appear in Work's children.
        XCTAssertTrue(work.children.contains { $0 === specs },
                      "Expected nb-specs in nb-work.children")
        XCTAssertEqual(1, work.children.count)
        // Leaf notebooks (no children) return an empty list.
        XCTAssertTrue(specs.children.isEmpty)
    }

    // MARK: - Notification posting

    func testAddNotebookPublishesNotebookAddedNotificationWhenHubWired() async throws {
        let repo = makeRepo()
        let hub = MessageHub()
        let dispatcher = ImmediateDispatcher.INSTANCE
        let notifHub = NotificationHub()
        defer { notifHub.dispose() }

        var observed: [VMx.Notification] = []
        var cancellables = Set<AnyCancellable>()
        notifHub.pending
            .sink { snapshot in
                for n in snapshot where !observed.contains(where: { $0 === n }) {
                    observed.append(n)
                }
            }
            .store(in: &cancellables)

        let vm = try NotebooksRootVM.builder()
            .name("root")
            .services(hub: hub, dispatcher: dispatcher)
            .repository(repo)
            .notificationHub(notifHub)
            .build()
        try vm.construct()
        try await vm.populate()

        try await vm.addNotebook(parentId: nil, name: "Inbox")

        // The notification posts via a detached Task (awaiting
        // NotificationHub.post would suspend until the toast resolves), so
        // wait briefly for its arrival.
        await waitUntil {
            observed.contains(where: { $0.message.contains("Notebook added") && $0.message.contains("Inbox") })
        }
        XCTAssertTrue(
            observed.contains(where: { $0.message.contains("Notebook added") && $0.message.contains("Inbox") }),
            "Expected 'Notebook added: \"Inbox\"' notification; got: \(observed.map(\.message))"
        )
    }

    func testAddNotebookWithParentIdAppearsUnderParentChildren() async throws {
        let repo = makeRepo()
        let vm = try buildVM(repo: repo)
        try vm.construct()
        try await vm.populate()

        guard let work = vm.all.first(where: { $0.model.id == "nb-work" }) else {
            return XCTFail("nb-work not found")
        }
        let beforeCount = work.children.count

        try await vm.addNotebook(parentId: "nb-work", name: "Subspecs")

        guard let added = vm.all.first(where: { $0.model.name == "Subspecs" }) else {
            return XCTFail("Subspecs not found after addNotebook")
        }
        XCTAssertEqual("nb-work", added.model.parentId)
        XCTAssertTrue(work.children.contains { $0 === added },
                      "Expected Subspecs in nb-work.children")
        XCTAssertEqual(beforeCount + 1, work.children.count)
    }

    // MARK: - PropertyChanged raises

    func testPopulateRaisesPropertyChangedForRoots() async throws {
        // Regression: without this raise, an already-bound view that's wired
        // before populate() completes never refreshes.
        let repo = makeRepo()
        let vm = try buildVM(repo: repo)
        try vm.construct()

        var raisedProps: [String] = []
        var cancellables = Set<AnyCancellable>()
        vm.hub.messages
            .compactMap { $0 as? PropertyChangedMessage }
            .filter { [weak vm] m in m.sender === vm }
            .sink { raisedProps.append($0.propertyName) }
            .store(in: &cancellables)

        try await vm.populate()

        XCTAssertTrue(raisedProps.contains("roots"),
                      "Expected 'roots' in raised property names; got \(raisedProps)")
    }

    func testAddNotebookRootLevelRaisesPropertyChangedForRoots() async throws {
        let repo = makeRepo()
        let vm = try buildVM(repo: repo)
        try vm.construct()
        try await vm.populate()

        var raisedProps: [String] = []
        var cancellables = Set<AnyCancellable>()
        vm.hub.messages
            .compactMap { $0 as? PropertyChangedMessage }
            .filter { [weak vm] m in m.sender === vm }
            .sink { raisedProps.append($0.propertyName) }
            .store(in: &cancellables)

        try await vm.addNotebook(parentId: nil, name: "Fresh Root")

        XCTAssertTrue(raisedProps.contains("roots"),
                      "Expected 'roots' in raised property names; got \(raisedProps)")
    }
}
